lib/remoteproc/elf_loader.c

```diff
@@ -579,10 +612,10 @@ int elf_load(struct remoteproc *rproc,
 		}
 		*nlen = nsize;
 		*nmemsize = nsegmsize;
-		phnums = elf_phnum(*img_info);
-		metal_log(METAL_LOG_DEBUG, "segment: %d, total segs %d\r\n",
-			  nsegment, phnums);
-		if (nsegment == phnums) {
```

Comment:
This is where it was comparing the last load `nsegment` against `phnums`, where `phnums` is the total number of program segments in the ELF.

Reply:
This looks okay to me. It's been more than 20 years since I last had a serious look at ELF file loading, so I had to do a little bit of refreshing. 

I think this is better than what we had before, but I'm still not sure it handles non-LOAD segments in the middle of the file. This might just be missing something.

Reply:
I think it can gracefully handle such non-LOAD segments. Because `nsegment` comes from `elf_next_load_segment()` that already has the capability to bypass non-LOAD segments and return `phdr` for the "next" LOAD segment that it finds (hence the function name). And the `max_load_phnum` parameter is also safe because there we don't count the total LOAD segments (there would have been problems if we were doing so) but rather find the segment number of the last LOAD segment in the ELF. So, I don't see intermediate non-LOAD segments creating any problem here.